### PR TITLE
Do not try to convert std::thread::id to integer value.

### PR DESCRIPTION
**Pull request by arrowd**
This is not allowed by the standard and libc++ is more stricter than libstdc++ on that matter.

Luckily, the conversion is used to turn it into a string, so just use stringstream to convert the thread::id directly.

Signed-off-by: Gleb Popov <6yearold@gmail.com>
Change-Id: Iea1a844a086b7fe7ed4703fd06e1d538d5d0bc43

* Target version: master 

### TODO

I didn't run-test this, as LOOL doesn't yet run on FreeBSD. This patch needs checking that I didn't broke anything.

### Checklist

- [X] Code is properly formatted
- [X] All commits have Change-Id
- [ ] I have run tests with `make check`
- [ ] I have issued `make run` and manually verified that everything looks okay
- [X] Documentation (manuals or wiki) has been updated or is not required

### CODE REVIEW

**Comment by Ashod** [comment]
Thanks for the patch @arrowd.
 
> This is not allowed by the standard and libc++ is more stricter than libstdc++ on that matter.

Do you have references please? 

> Luckily, the conversion is used to turn it into a string, so just use stringstream to convert the thread::id directly.

Well, your patch changes serialization of long to string. I don't see he how that is related to thread::id at all.

File: common/Log.cpp [hunk 1]
```diff
@@ -194,22 +194,11 @@ namespace Log
...
-        }
-        else
-        {
-            to_ascii_fixed<5>(pos, osTid);
-            pos += 5;
-        }
+        auto osTid = Util::getThreadId();
+        std::stringstream ss;
+        ss << osTid;
+        pos = strcopy(ss.str().c_str(), pos);
```
Comment (Ashod) [inline common/Log.cpp:200]: What is "strcopy" ? Are you sure this code compiles?

**Review by Ashod — CHANGES_REQUESTED** [review]

**Comment by arrowd** [comment]
> Do you have references please? 
> Well, your patch changes serialization of long to string. I don't see he how that is related to thread::id at all.

The long is first initialized with thread::id(), which is not allowed according to https://stackoverflow.com/q/7432100/637669
The `thread::id` definition doesn't say anything about its implementation, so we can't assume that it can be converted into an integer value: https://en.cppreference.com/w/cpp/thread/thread/id

File: common/Log.cpp (hunk 1 above)
Comment (arrowd) [reply common/Log.cpp:200]: `strcopy` is also used below, on line 232. It is defined in the same file on line 153/

Of course I checked that the code compiles. It wasn't compiling for before this change.

**Comment by Ashod** [comment]
> > Do you have references please?
> > Well, your patch changes serialization of long to string. I don't see he how that is related to thread::id at all.
> 
> The long is first initialized with thread::id(), which is not allowed according to https://stackoverflow.com/q/7432100/637669
> The `thread::id` definition doesn't say anything about its implementation, so we can't assume that it can be converted into an integer value: https://en.cppreference.com/w/cpp/thread/thread/id

Thanks, but as I already said, your patch doesn't do anything at all with thread::id. It only changes the conversion from long to string (by making it slower). I don't see how this solves anything related to thread::id at all. Are you saying the old code doesn't compile on FreeBSD and it now works? Have you compiled on **FreeBSD**?

**Comment by arrowd** [comment]
> Thanks, but as I already said, your patch doesn't do anything at all with thread::id. It only changes the conversion from long to string (by making it slower). I don't see how this solves anything related to thread::id at all. Are you saying the old code doesn't compile on FreeBSD and it now works? Have you compiled on **FreeBSD**?

As I said in the review for line 200:

> Of course I checked that the code compiles. It wasn't compiling for FreeBSD before this change.

The old code also did convert the id into string, but was using `to_ascii{_fixed}` functions for that.

**Review by tml1024 — APPROVED** [review]
Looks fine to me. It is not uncommon for our code to (accidentally) use features specific to libstdc++, as I have noticed myself when building it with libc++ (for iOS). It is good if building with libc++ on other platforms, too, forces us to be more standard compliant.

**Comment by mikekaganski** [comment]
> Thanks, but as I already said, your patch doesn't do anything at all with thread::id. It only changes the conversion from long to string (by making it slower).

The patch changes

```c++
const long osTid = Util::getThreadId();
```

into

```c++
auto osTid = Util::getThreadId();
```

Given that common/Util.hpp declares `getThreadId` as

```c++
#ifdef __linux
    pid_t getThreadId();
#else
    std::thread::id getThreadId();
#endif
```

it's not convert `long`, but a platform-dependent type, which is likely not `long` on *BSD (which @arrowd implements support for).

**Comment by Ashod** [comment]
> Looks fine to me. It is not uncommon for our code to (accidentally) use features specific to libstdc++, as I have noticed myself when building it with libc++ (for iOS). It is good if building with libc++ on other platforms, too, forces us to be more standard compliant.

I don't think that's what's happening though. I had to check the code to understand why you (@tml1024) and the original author (@arrowd) seem to think we are misusing the standard.

Looking at Util::getThreadId makes clear what's happening. For Linux we actually use syscall(SYS_gettid); and return pid_t. This is a perfect integer and the code is 100% correct in treating it as an integer and doing the conversion to string.

Of course we also provide a fallback for Util::getThreadId() for non-Linux systems, which returns std::hread::id. This wouldn't assign to long and that simply doesn't compile. This seems to be what @arrowd describing as "didn't compile" on FreeBSD.

So at least in this case, we aren't using any "features specific to libstdc++," either accidentally or intentionally. That's the good news. The bad news is that using std::stringstream for the conversion on each and every log entry will be painfully slow.

@arrowd I now understand what your patch is trying to do and what problem it is trying to solve. Sorry that it wasn't clear to me and I was trying to clarify. Now, what I suggest is to specialize your new code for non __linux systems:

```
        // Thread ID.
        const auto osTid = Util::getThreadId();
#ifdef __linux
        // On Linux osTid is pid_t.
        if (osTid > 99999)
        {
            if (osTid > 999999)
                pos = to_ascii(pos, osTid);
            else
            {
                to_ascii_fixed<6>(pos, osTid);
                pos += 6;
            }
        }
        else
        {
            to_ascii_fixed<5>(pos, osTid);
            pos += 5;
        }
#else
        // On all other systems osTid is std::thread::id.
        std::stringstream ss;
        ss << osTid;
        pos = strcopy(ss.str().c_str(), pos);
#endif
```

This code gets executed on each and every log entry, and it has to be fast (and safe). I think the above is the right balance as it treats the two types (pid_t and std::thread::id) differently and efficiently.

**Comment by Ashod** [comment]
> it's not convert `long`, but a platform-dependent type, which is likely not `long` on *BSD (which @arrowd implements support for).

Thanks @mikekaganski, I have already realized that the problem is that we were solving a special case by changing the code for the most common case (Linux). I made my suggestion in the patch.

**Comment by tml1024** [comment]
> This code gets executed on each and every log entry, and it has to be fast (and safe). 

But surely we shouldn't just assume that something has any signifiant performance impact without, you know, measurements? Premature optimisation and all that. Having code that is simple and correct is better than having code that is more complicated and possibly a tiny bit faster. In my humble opinion.

**Comment by Ashod** [comment]
@arrowd, please checkout this PR and build it on FreeBSD. I've added a commit to special-case for Linux. I can build both cases on Linux without problems, but would be good to get confirmation on FreeBSD before we merge this.

Once you confirm, I'm happy to merge.

Thanks again for your contribution!

**Comment by mmeeks** [comment]
It is easy to see logging in the profiles, and things crawling when it is enabled. Then again - if we think the thread-id-to-string conversion is particularly bad, we could easily cache the result in a thread-local variable (I guess) =)

**Comment by arrowd** [comment]
> @arrowd, please checkout this PR and build it on FreeBSD. I've added a commit to special-case for Linux. I can build both cases on Linux without problems, but would be good to get confirmation on FreeBSD before we merge this.
> 
> Once you confirm, I'm happy to merge.

Yes, your change also works for me.

> 
> Thanks again for your contribution!

And thank you for reworking my patch. I completely missed the fact that `Util::getThreadId()` have different return types depending on the OS.

**Comment by Ashod** [comment]
> Yes, your change also works for me.

Great, thanks for the confirmation!

> > Thanks again for your contribution!
> 
> And thank you for reworking my patch. I completely missed the fact that `Util::getThreadId()` have different return types depending on the OS.

Funny thing is that I also forgot that the we returned two different types from `Util::getThreadId()`! :) And this is why I was confused at first. But happy we figured it out and have something solid now that works for everyone.

Btw, I didn't rework your patch, it remains intact. I only added a second commit. So your patch will go in as it's own commit, exactly as you authored it, followed with my patch to special-case for Linux. (You can see the two commits from the "Commits" tab above.)

Thanks again Gleb, this is a really good patch!

**Review by Ashod — APPROVED** [review]

**Pull request merged**